File: matlab/kmzwrite.m
function kmzwrite(kmzPath, varargin)
%KMZWRITE  Write a KMZ containing a raster surface, contour lines, coloured
%          points and marker points. Requires no toolboxes.
%
%   KMZWRITE(FILE, 'Name', Value, ...) writes a Google Earth KMZ to FILE.
%   Every section is optional; supply only what you have.
%
%   Surface (all three required together):
%     'GridLat'          Matrix of latitudes, as produced by meshgrid
%     'GridLon'          Matrix of longitudes, same size
%     'GridValue'        Matrix of values, same size. NaN cells are written
%                        transparent, so a griddata result is fine as-is.
%     'SurfaceOpacity'   0..1, default 0.8
%     'ContourLevels'    Number of contour lines, default 10. Use 0 to omit
%                        them. They go in their own folder, switched off by
%                        default so the raster reads cleanly.
%
%   Coloured points:
%     'PointLat', 'PointLon'   Vectors of equal length
%     'PointAlt'               Vector of altitudes, default 0
%     'PointValue'             Vector driving the marker colour, default []
%     'PointAltitudeMode'      'absolute' | 'relativeToGround' |
%                              'clampToGround'. Default 'clampToGround'
%
%   Markers (e.g. a known tag position):
%     'MarkerLat', 'MarkerLon' Vectors of equal length
%     'MarkerName'             Char, or cellstr with one name per marker
%
%   Folder labels shown in the Google Earth tree:
%     'PointFolderName'        Default 'Points'
%     'MarkerFolderName'       Default 'Markers'
%
%   Labelling:
%     'Name'         Document name, default the output file name
%     'Description'  Document description
%     'ValueName'    What the values are, used in the legend and balloons.
%                    Default 'Value'
%
%   Example:
%     [X,Y]   = meshgrid(xVec, yVec);
%     [LAT,LON] = enuToGeodetic(X, Y, home);
%     kmzwrite('scan.kmz', 'GridLat',LAT, 'GridLon',LON, ...
%              'GridValue',snrGrid, 'ValueName','SNR (dB)', ...
%              'MarkerLat',tagLat, 'MarkerLon',tagLon, 'MarkerName','Tag');
%
%   Written to replace the kmltoolbox calls this project used to make. That
%   toolbox inlined a full <Style> block into every placemark and pointed
%   each marker icon at http://maps.google.com, so exports were roughly
%   three times larger than necessary and did not render without a network.
%   Here styles are declared once and referenced by styleUrl, and the marker
%   icon is packaged inside the kmz.

    %% ---- options ----------------------------------------------------
    [~, defaultName] = fileparts(kmzPath);
    opt = struct( ...
        'GridLat',           [], ...
        'GridLon',           [], ...
        'GridValue',         [], ...
        'SurfaceOpacity',    0.8, ...
        'ContourLevels',     10, ...
        'PointLat',          [], ...
        'PointLon',          [], ...
        'PointAlt',          [], ...
        'PointValue',        [], ...
        'PointAltitudeMode', 'clampToGround', ...
        'MarkerLat',         [], ...
        'MarkerLon',         [], ...
        'MarkerName',        'Marker', ...
        'PointFolderName',   'Points', ...
        'MarkerFolderName',  'Markers', ...
        'Name',              defaultName, ...
        'Description',       '', ...
        'ValueName',         'Value');

    if mod(numel(varargin), 2) ~= 0
        error('kmzwrite:badPairs', 'Options must be name/value pairs.');
    end
    known = fieldnames(opt);
    for i = 1:2:numel(varargin)
        name = varargin{i};
        if ~(ischar(name) || isstring(name))
            error('kmzwrite:badName', 'Option %d is not a name.', (i+1)/2);
        end
        hit = strcmpi(known, name);
        if ~any(hit)
            error('kmzwrite:unknownOption', 'Unknown option "%s".', char(name));
        end
        opt.(known{hit}) = varargin{i+1};
    end

    nBins    = 64;      % colour bins for points
    nLineCol = 16;      % colour bins for contour lines
    cmap     = turbo(nBins);

    haveGrid = ~isempty(opt.GridValue) && ~isempty(opt.GridLat) && ...
               ~isempty(opt.GridLon)   && any(isfinite(opt.GridValue(:)));
    if haveGrid && ~isequal(size(opt.GridLat), size(opt.GridLon), size(opt.GridValue))
        error('kmzwrite:gridSize', 'GridLat, GridLon and GridValue must be the same size.');
    end

    pLat = opt.PointLat(:);
    pLon = opt.PointLon(:);
    if numel(pLat) ~= numel(pLon)
        error('kmzwrite:pointSize', 'PointLat and PointLon must be the same length.');
    end
    nPts = numel(pLat);
    pAlt = localFill(opt.PointAlt(:), nPts, 0);
    pVal = opt.PointValue(:);
    if ~isempty(pVal) && numel(pVal) ~= nPts
        error('kmzwrite:pointSize', 'PointValue must match PointLat in length.');
    end

    mLat = opt.MarkerLat(:);
    mLon = opt.MarkerLon(:);
    if numel(mLat) ~= numel(mLon)
        error('kmzwrite:markerSize', 'MarkerLat and MarkerLon must be the same length.');
    end
    mName = opt.MarkerName;
    if ischar(mName) || isstring(mName)
        mName = repmat({char(mName)}, numel(mLat), 1);
    end

    %% ---- staging -----------------------------------------------------
    stage = [tempname, '_kmz'];
    mkdir(stage);
    mkdir(fullfile(stage, 'files'));
    cleanupStage = onCleanup(@() rmdir(stage, 's')); %#ok<NASGU>

    parts = {};
    parts{end+1} = sprintf('<?xml version="1.0" encoding="UTF-8"?>\n');
    parts{end+1} = sprintf('<kml xmlns="http://www.opengis.net/kml/2.2">\n<Document>\n');
    parts{end+1} = sprintf('<name>%s</name>\n', localEscape(opt.Name));
    if ~isempty(opt.Description)
        parts{end+1} = sprintf('<description>%s</description>\n', localEscape(opt.Description));
    end

    %% ---- shared styles ------------------------------------------------
    localWriteDotPng(fullfile(stage, 'files', 'dot.png'));
    for k = 1:nBins
        parts{end+1} = sprintf(['<Style id="p%02d"><IconStyle><color>%s</color>' ...
            '<scale>0.5</scale><Icon><href>files/dot.png</href></Icon></IconStyle>' ...
            '<LabelStyle><scale>0</scale></LabelStyle></Style>\n'], ...
            k-1, localKmlColor(cmap(k,:), 255)); %#ok<AGROW>
    end
    lineCmap = turbo(nLineCol);
    for k = 1:nLineCol
        parts{end+1} = sprintf(['<Style id="l%02d"><LineStyle><color>%s</color>' ...
            '<width>2</width></LineStyle></Style>\n'], ...
            k-1, localKmlColor(lineCmap(k,:), 255)); %#ok<AGROW>
    end
    parts{end+1} = sprintf(['<Style id="marker"><IconStyle><color>FF000000</color>' ...
        '<scale>1.4</scale><Icon><href>files/dot.png</href></Icon></IconStyle>' ...
        '</Style>\n']);

    %% ---- ground overlay ------------------------------------------------
    if haveGrid
        localWriteSurfacePng(fullfile(stage, 'files', 'surface.png'), opt.GridValue);

        LAT = opt.GridLat;
        LON = opt.GridLon;
        % GroundOverlay maps image edges to the box, but grid values sit at
        % cell centres, so the box grows by half a cell in each direction.
        halfLat = 0;
        halfLon = 0;
        if size(LAT, 1) > 1, halfLat = abs(LAT(2,1) - LAT(1,1))/2; end
        if size(LON, 2) > 1, halfLon = abs(LON(1,2) - LON(1,1))/2; end

        alphaHex = dec2hex(uint8(round(255*min(max(opt.SurfaceOpacity, 0), 1))), 2);
        parts{end+1} = sprintf(['<Folder><name>%s surface</name>\n' ...
            '<GroundOverlay><name>%s</name><color>%sffffff</color>\n' ...
            '<Icon><href>files/surface.png</href></Icon>\n' ...
            '<LatLonBox><north>%.10f</north><south>%.10f</south>' ...
            '<east>%.10f</east><west>%.10f</west></LatLonBox>\n' ...
            '</GroundOverlay></Folder>\n'], ...
            localEscape(opt.ValueName), localEscape(opt.ValueName), lower(alphaHex), ...
            max(LAT(:)) + halfLat, min(LAT(:)) - halfLat, ...
            max(LON(:)) + halfLon, min(LON(:)) - halfLon);
    end

    %% ---- contour lines, off by default ---------------------------------
    if haveGrid && opt.ContourLevels > 0
        lonVec = opt.GridLon(1,:);
        latVec = opt.GridLat(:,1).';
        C = contourc(lonVec, latVec, opt.GridValue, opt.ContourLevels);

        finiteG = opt.GridValue(isfinite(opt.GridValue));
        gLo = min(finiteG);
        gHi = max(finiteG);
        if gHi <= gLo, gHi = gLo + 1; end

        segs = {};
        k = 1;
        while k < size(C, 2)
            lev = C(1, k);
            np  = C(2, k);
            if np < 2 || k + np > size(C, 2)
                break
            end
            xs = C(1, k+1:k+np);
            ys = C(2, k+1:k+np);
            ci = min(max(round((lev - gLo)/(gHi - gLo)*(nLineCol - 1)) + 1, 1), nLineCol);
            segs{end+1} = sprintf(['<Placemark><name>%.4g</name>' ...
                '<styleUrl>#l%02d</styleUrl><LineString><tessellate>1</tessellate>' ...
                '<altitudeMode>clampToGround</altitudeMode>' ...
                '<coordinates>%s</coordinates></LineString></Placemark>\n'], ...
                lev, ci-1, sprintf('%.8f,%.8f,0 ', [xs; ys])); %#ok<AGROW>
            k = k + np + 1;
        end

        if ~isempty(segs)
            parts{end+1} = sprintf('<Folder><name>Contour lines</name><visibility>0</visibility>\n');
            parts{end+1} = strjoin(segs, '');
            parts{end+1} = sprintf('</Folder>\n');
        end
    end

    %% ---- coloured points -------------------------------------------------
    if nPts > 0
        if isempty(pVal)
            binIdx = repmat(nBins, 1, nPts);
            valTxt = repmat({''}, nPts, 1);
        else
            finiteP = pVal(isfinite(pVal));
            if isempty(finiteP) || numel(unique(finiteP)) == 1
                binIdx = repmat(nBins, 1, nPts);
            else
                vLo = min(finiteP);
                vHi = max(finiteP);
                binIdx = round((pVal(:).' - vLo)/(vHi - vLo)*(nBins - 1)) + 1;
                binIdx(~isfinite(binIdx)) = 1;
                binIdx = min(max(binIdx, 1), nBins);
            end
        end

        % The value name goes into the format string, so escape any percent.
        label = strrep(localEscape(opt.ValueName), '%', '%%');
        mode  = localEscape(opt.PointAltitudeMode);
        if isempty(pVal)
            fmt = ['<Placemark><styleUrl>#p%02d</styleUrl>' ...
                   '<Point><altitudeMode>' mode '</altitudeMode>' ...
                   '<coordinates>%.8f,%.8f,%.2f</coordinates></Point></Placemark>' char(10)];
            block = sprintf(fmt, [binIdx - 1; pLon(:).'; pLat(:).'; pAlt(:).']);
        else
            fmt = ['<Placemark><styleUrl>#p%02d</styleUrl>' ...
                   '<description>' label ' = %.4g</description>' ...
                   '<Point><altitudeMode>' mode '</altitudeMode>' ...
                   '<coordinates>%.8f,%.8f,%.2f</coordinates></Point></Placemark>' char(10)];
            block = sprintf(fmt, [binIdx - 1; pVal(:).'; pLon(:).'; pLat(:).'; pAlt(:).']);
        end
        parts{end+1} = sprintf('<Folder><name>%s</name>\n', localEscape(opt.PointFolderName));
        parts{end+1} = block;
        parts{end+1} = sprintf('</Folder>\n');
    end

    %% ---- markers ----------------------------------------------------------
    if ~isempty(mLat)
        parts{end+1} = sprintf('<Folder><name>%s</name>\n', localEscape(opt.MarkerFolderName));
        for i = 1:numel(mLat)
            parts{end+1} = sprintf(['<Placemark><name>%s</name><styleUrl>#marker</styleUrl>' ...
                '<Point><altitudeMode>clampToGround</altitudeMode>' ...
                '<coordinates>%.8f,%.8f,0</coordinates></Point></Placemark>\n'], ...
                localEscape(mName{i}), mLon(i), mLat(i)); %#ok<AGROW>
        end
        parts{end+1} = sprintf('</Folder>\n');
    end

    parts{end+1} = sprintf('</Document>\n</kml>\n');

    %% ---- write and zip -----------------------------------------------------
    kmlFile = fullfile(stage, 'doc.kml');
    fid = fopen(kmlFile, 'w');
    if fid < 0
        error('kmzwrite:open', 'Could not open %s for writing.', kmlFile);
    end
    fwrite(fid, unicode2native(strjoin(parts, ''), 'UTF-8'));
    fclose(fid);

    zipPath = [tempname, '.zip'];
    zip(zipPath, {'doc.kml', 'files'}, stage);
    if ~isfile(zipPath) && isfile([zipPath, '.zip'])
        zipPath = [zipPath, '.zip'];
    end
    movefile(zipPath, kmzPath, 'f');
end


function v = localFill(v, n, defaultValue)
    if isempty(v)
        v = repmat(defaultValue, n, 1);
    elseif isscalar(v)
        v = repmat(v, n, 1);
    elseif numel(v) ~= n
        error('kmzwrite:vectorSize', 'Expected %d elements, got %d.', n, numel(v));
    end
end


function out = localEscape(txt)
    out = char(txt);
    out = strrep(out, '&', '&amp;');
    out = strrep(out, '<', '&lt;');
    out = strrep(out, '>', '&gt;');
end


function hex = localKmlColor(rgbRow, alpha)
    % KML colours are aabbggrr, the reverse of the usual order.
    rgbRow = min(max(rgbRow(:).', 0), 1);
    v = uint8(round([alpha, 255*rgbRow(3), 255*rgbRow(2), 255*rgbRow(1)]));
    hex = sprintf('%02X%02X%02X%02X', v(1), v(2), v(3), v(4));
end


function localWriteDotPng(pngPath)
    % A white disc with a soft edge. White so the per-style <color> tint in
    % IconStyle reproduces the intended colour exactly.
    m = 64;
    [xx, yy] = meshgrid(linspace(-1, 1, m));
    r = hypot(xx, yy);
    alpha = min(max((0.90 - r)*(m/6), 0), 1);
    imwrite(ones(m, m, 3), pngPath, 'Alpha', alpha);
end


function localWriteSurfacePng(pngPath, G)
    % PNG row 1 is the north edge of a GroundOverlay, but row 1 of a meshgrid
    % is its south edge, hence the flip. NaN cells are written transparent.
    finiteMask = isfinite(G);
    lo = min(G(finiteMask));
    hi = max(G(finiteMask));
    if isempty(lo) || hi <= lo
        hi = lo + 1;
    end
    idx = round((G - lo)/(hi - lo)*255) + 1;
    idx(~finiteMask) = 1;
    idx = min(max(idx, 1), 256);

    cmap = turbo(256);
    rgb  = reshape(cmap(idx(:), :), [size(G), 3]);
    imwrite(flipud(rgb), pngPath, 'Alpha', flipud(double(finiteMask)));
end

File: python/kmzwrite.py
"""KMZ writer. Port of kmzwrite.m.

Writes a Google Earth KMZ holding an optional raster surface, contour lines,
colour-coded points and marker points. Styles are declared once and referenced
by styleUrl, and the marker icon is packaged inside the archive, so exports are
compact and render with no network.
"""

import io
import os
import zipfile

import numpy as np
from matplotlib import colormaps
import matplotlib.pyplot as plt

N_POINT_BINS = 64
N_LINE_BINS = 16


def _escape(text):
    return (str(text).replace("&", "&amp;")
                     .replace("<", "&lt;")
                     .replace(">", "&gt;"))


def _kml_color(rgb, alpha=255):
    """KML colours are aabbggrr, the reverse of the usual order."""
    r, g, b = [int(round(255 * min(max(float(c), 0.0), 1.0))) for c in rgb[:3]]
    return "%02X%02X%02X%02X" % (int(alpha), b, g, r)


def _dot_png():
    """A white disc with a soft edge, tinted per style by IconStyle <color>."""
    m = 64
    xx, yy = np.meshgrid(np.linspace(-1, 1, m), np.linspace(-1, 1, m))
    r = np.hypot(xx, yy)
    alpha = np.clip((0.90 - r) * (m / 6.0), 0.0, 1.0)
    rgba = np.dstack([np.ones((m, m)), np.ones((m, m)), np.ones((m, m)), alpha])
    buf = io.BytesIO()
    plt.imsave(buf, rgba, format="png")
    return buf.getvalue()


def _surface_png(grid):
    """Turbo-mapped raster, transparent where the grid is NaN.

    PNG row 0 is the north edge of a GroundOverlay but row 0 of a meshgrid is
    its south edge, hence the flip.
    """
    grid = np.asarray(grid, dtype=float)
    finite = np.isfinite(grid)
    lo = float(grid[finite].min())
    hi = float(grid[finite].max())
    if hi <= lo:
        hi = lo + 1.0
    norm = np.clip((grid - lo) / (hi - lo), 0.0, 1.0)
    norm[~finite] = 0.0
    rgba = colormaps["turbo"](norm)
    rgba[..., 3] = finite.astype(float)
    buf = io.BytesIO()
    plt.imsave(buf, np.flipud(rgba), format="png")
    return buf.getvalue()


def kmzwrite(kmz_path,
             grid_lat=None, grid_lon=None, grid_value=None,
             surface_opacity=0.8, contour_levels=10,
             point_lat=None, point_lon=None, point_alt=None, point_value=None,
             point_altitude_mode="clampToGround",
             marker_lat=None, marker_lon=None, marker_name="Marker",
             point_folder_name="Points", marker_folder_name="Markers",
             name=None, description="", value_name="Value"):
    """Write a KMZ. Every section is optional; supply only what you have."""

    if name is None:
        name = os.path.splitext(os.path.basename(kmz_path))[0]

    have_grid = (grid_value is not None and grid_lat is not None
                 and grid_lon is not None
                 and np.any(np.isfinite(np.asarray(grid_value, dtype=float))))
    if have_grid:
        grid_lat = np.asarray(grid_lat, dtype=float)
        grid_lon = np.asarray(grid_lon, dtype=float)
        grid_value = np.asarray(grid_value, dtype=float)
        if not (grid_lat.shape == grid_lon.shape == grid_value.shape):
            raise ValueError("grid_lat, grid_lon and grid_value must match in shape")

    p_lat = np.atleast_1d(np.asarray(point_lat, dtype=float)) if point_lat is not None else np.empty(0)
    p_lon = np.atleast_1d(np.asarray(point_lon, dtype=float)) if point_lon is not None else np.empty(0)
    if p_lat.size != p_lon.size:
        raise ValueError("point_lat and point_lon must be the same length")
    n_pts = p_lat.size
    if point_alt is None:
        p_alt = np.zeros(n_pts)
    else:
        p_alt = np.atleast_1d(np.asarray(point_alt, dtype=float))
        if p_alt.size == 1:
            p_alt = np.full(n_pts, float(p_alt[0]))
    p_val = np.atleast_1d(np.asarray(point_value, dtype=float)) if point_value is not None else None
    if p_val is not None and p_val.size != n_pts:
        raise ValueError("point_value must match point_lat in length")

    m_lat = np.atleast_1d(np.asarray(marker_lat, dtype=float)) if marker_lat is not None else np.empty(0)
    m_lon = np.atleast_1d(np.asarray(marker_lon, dtype=float)) if marker_lon is not None else np.empty(0)
    if m_lat.size != m_lon.size:
        raise ValueError("marker_lat and marker_lon must be the same length")
    if isinstance(marker_name, str):
        m_names = [marker_name] * m_lat.size
    else:
        m_names = list(marker_name)

    parts = ['<?xml version="1.0" encoding="UTF-8"?>\n',
             '<kml xmlns="http://www.opengis.net/kml/2.2">\n<Document>\n',
             "<name>%s</name>\n" % _escape(name)]
    if description:
        parts.append("<description>%s</description>\n" % _escape(description))

    # ---- shared styles ------------------------------------------------
    point_cmap = colormaps["turbo"].resampled(N_POINT_BINS)
    for k in range(N_POINT_BINS):
        parts.append(
            '<Style id="p%02d"><IconStyle><color>%s</color><scale>0.5</scale>'
            '<Icon><href>files/dot.png</href></Icon></IconStyle>'
            '<LabelStyle><scale>0</scale></LabelStyle></Style>\n'
            % (k, _kml_color(point_cmap(k))))
    line_cmap = colormaps["turbo"].resampled(N_LINE_BINS)
    for k in range(N_LINE_BINS):
        parts.append(
            '<Style id="l%02d"><LineStyle><color>%s</color><width>2</width>'
            '</LineStyle></Style>\n' % (k, _kml_color(line_cmap(k))))
    parts.append('<Style id="marker"><IconStyle><color>FF000000</color>'
                 '<scale>1.4</scale><Icon><href>files/dot.png</href></Icon>'
                 '</IconStyle></Style>\n')

    files = {"files/dot.png": _dot_png()}

    # ---- ground overlay -----------------------------------------------
    if have_grid:
        files["files/surface.png"] = _surface_png(grid_value)
        # GroundOverlay maps image edges to the box, but grid values sit at
        # cell centres, so the box grows by half a cell each way.
        half_lat = abs(grid_lat[1, 0] - grid_lat[0, 0]) / 2 if grid_lat.shape[0] > 1 else 0.0
        half_lon = abs(grid_lon[0, 1] - grid_lon[0, 0]) / 2 if grid_lon.shape[1] > 1 else 0.0
        alpha_hex = "%02x" % int(round(255 * min(max(surface_opacity, 0.0), 1.0)))
        parts.append(
            '<Folder><name>%s surface</name>\n<GroundOverlay><name>%s</name>'
            '<color>%sffffff</color>\n'
            '<Icon><href>files/surface.png</href></Icon>\n'
            '<LatLonBox><north>%.10f</north><south>%.10f</south>'
            '<east>%.10f</east><west>%.10f</west></LatLonBox>\n'
            '</GroundOverlay></Folder>\n'
            % (_escape(value_name), _escape(value_name), alpha_hex,
               grid_lat.max() + half_lat, grid_lat.min() - half_lat,
               grid_lon.max() + half_lon, grid_lon.min() - half_lon))

    # ---- contour lines, off by default --------------------------------
    if have_grid and contour_levels and contour_levels > 0:
        finite = grid_value[np.isfinite(grid_value)]
        g_lo, g_hi = float(finite.min()), float(finite.max())
        if g_hi <= g_lo:
            g_hi = g_lo + 1.0
        fig = plt.figure()
        try:
            cs = plt.contour(grid_lon, grid_lat, grid_value, levels=contour_levels)
            segments = []
            for level, paths in zip(cs.levels, cs.allsegs):
                ci = int(np.clip(round((level - g_lo) / (g_hi - g_lo) * (N_LINE_BINS - 1)),
                                 0, N_LINE_BINS - 1))
                for path in paths:
                    if len(path) < 2:
                        continue
                    coords = " ".join("%.8f,%.8f,0" % (x, y) for x, y in path)
                    segments.append(
                        '<Placemark><name>%.4g</name><styleUrl>#l%02d</styleUrl>'
                        '<LineString><tessellate>1</tessellate>'
                        '<altitudeMode>clampToGround</altitudeMode>'
                        '<coordinates>%s</coordinates></LineString></Placemark>\n'
                        % (level, ci, coords))
        finally:
            plt.close(fig)
        if segments:
            parts.append('<Folder><name>Contour lines</name>'
                         '<visibility>0</visibility>\n')
            parts.extend(segments)
            parts.append("</Folder>\n")

    # ---- points --------------------------------------------------------
    if n_pts:
        if p_val is None:
            bins = np.full(n_pts, N_POINT_BINS - 1, dtype=int)
        else:
            finite = p_val[np.isfinite(p_val)]
            if finite.size == 0 or np.unique(finite).size == 1:
                bins = np.full(n_pts, N_POINT_BINS - 1, dtype=int)
            else:
                lo, hi = finite.min(), finite.max()
                idx = np.round((p_val - lo) / (hi - lo) * (N_POINT_BINS - 1))
                idx[~np.isfinite(idx)] = 0
                bins = np.clip(idx, 0, N_POINT_BINS - 1).astype(int)

        label = _escape(value_name)
        mode = _escape(point_altitude_mode)
        parts.append("<Folder><name>%s</name>\n" % _escape(point_folder_name))
        chunks = []
        for i in range(n_pts):
            if p_val is None:
                desc = ""
            else:
                desc = "<description>%s = %.4g</description>" % (label, p_val[i])
            chunks.append(
                '<Placemark><styleUrl>#p%02d</styleUrl>%s'
                '<Point><altitudeMode>%s</altitudeMode>'
                '<coordinates>%.8f,%.8f,%.2f</coordinates></Point></Placemark>\n'
                % (bins[i], desc, mode, p_lon[i], p_lat[i], p_alt[i]))
        parts.append("".join(chunks))
        parts.append("</Folder>\n")

    # ---- markers -------------------------------------------------------
    if m_lat.size:
        parts.append("<Folder><name>%s</name>\n" % _escape(marker_folder_name))
        for i in range(m_lat.size):
            parts.append(
                '<Placemark><name>%s</name><styleUrl>#marker</styleUrl>'
                '<Point><altitudeMode>clampToGround</altitudeMode>'
                '<coordinates>%.8f,%.8f,0</coordinates></Point></Placemark>\n'
                % (_escape(m_names[i]), m_lon[i], m_lat[i]))
        parts.append("</Folder>\n")

    parts.append("</Document>\n</kml>\n")

    with zipfile.ZipFile(kmz_path, "w", zipfile.ZIP_DEFLATED) as z:
        z.writestr("doc.kml", "".join(parts))
        for arcname, blob in files.items():
            z.writestr(arcname, blob)
    return kmz_path
